grayscale.py

```diff
@@ -1,764 +1,549 @@
-        # desired result is achieved anyway. To simplify things, the following
-        # comments are written as if the assumptions _are_ correct.
-
-        # We will keep the display synchronised by resetting the row counter
-        # before each frame and then outputting a frame of 57 rows. This is 17
-        # rows past the 40 of the actual display.
-
-        # Prior to loading in the frame we park the row counter at row 0 and
-        # wait for the nominal time for 8 rows to be output. This (hopefully)
-        # provides enough time for the row counter to reach row 0 before it
-        # sticks there. (Note: recent test indicate that perhaps the current row
-        # actually jumps before parking)
-        # The 'parking' is done by setting the number of rows (aka 'multiplex
-        # ratio') to 1 row. This is an invalid setting according to the datasheet
-        # but seems to still have the desired effect.
+        # Public and compatibility variables
+        self.width = _WIDTH
+        self.height = _HEIGHT
+        self.max_x = _WIDTH - 1
+        self.max_y = _HEIGHT - 1
+        self.lastUpdateEnd = 0
+        self.frameRate = 0
+        self.display = self # This acts as both the GraphicsClass and SSD1306
+        self.contrast = self.brightness
+
+        # Draw buffers.
+        # This comprises of two full buffer lengths.
+        # The first section contains black and white compatible
+        # with the display buffer from the standard Thumby API,
+        # and the second contains the shading to create
+        # offwhite (lightgray) or offblack (darkgray).
+        self._drawBuffer = bytearray(_BUFF_SIZE*2)
+        # The base "buffer" matches compatibility with the std Thumby API.
+        self.buffer = memoryview(self._drawBuffer)[:_BUFF_SIZE]
+        # The "shading" buffer adds the grayscale
+        self.shading = memoryview(self._drawBuffer)[_BUFF_SIZE:]
+
+        # Display device configuration
+        self._spi = SPI(0, sck=Pin(18), mosi=Pin(19))
+        self._dc = Pin(17)
+        self._cs = Pin(16)
+        self._res = Pin(20)
+        self._res.init(Pin.OUT, value=1)
+        self._spi.init(baudrate=100 * 1000 * 1000, polarity=0, phase=0)
+        # Black and White direct display for when the GPU is off (also inits)
+        self._displayBW = SSD1306_SPI(_WIDTH, _HEIGHT,
+            self._spi, dc=self._dc, res=self._res, cs=self._cs)
+        # Inject buffer to Black and White display
+        self._displayBW.buffer = self.buffer
+
+        # Display driver subframe buffers.
+        # These essentially combine into one framebuffer,
+        # but are split into three buffers that hit the display
+        # at different subframe intervals.
+        self._subframes = array('O', [bytearray(_BUFF_SIZE),
+            bytearray(_BUFF_SIZE), bytearray(_BUFF_SIZE)])
+
+        # Timing and display synchronisation commands.
+        #
+        # DISCLAIMER: The following comments relating to the hardware
```

Comment:
Mostly it was trying to fit to 72 cols for comment blocks. The bit I trimmed, I arrogantly felt it slowed down my rate of understanding the code, but didn't leave me any wiser. But you are the boss/expert with this stuff! Happy to add it back and sorry for the noise! Pushed.

Reply:
Ah ok. I tend to leave existing comments alone unless the code they refer to has changed and made them incorrect.

FWIW I think @Timendus was using 80 columns... although I think I probably did go past 80 cols in some places.

It might be worth do some interactive rebasing with what you currently have as these revert commits aren't really reverting, they're just re-editing to what the previous content was. Or to put it another way, git-blame won't be accurate anymore.

Reply:
Ah, yeah. I'm rushing forwards. :(

Reply:
I'm just going to do this stuff quickly Someone else should apply the changes and this branch should be killed. I'm going to keep it as draft. It feels like this PR is long dead in terms of merge viability anyone so I'm deciding it will stay a draft forever and someone else can take the "blame" so to speak, by applying a raw diff of whatever they want.

Reply:
FWIW I think you should definitely be involved in creating the PRs that capture the relevant changes taken from here. In a positive-sense, you should take the blame! :-)